tests/test_server.py

```diff
@@ -38,204 +53,206 @@ def test_home(client):
-        ['/api/bookmarks/refresh', response_template['failure']]
+    'method, url, exp_res, data', [
+        ['get', '/api/tags', Response.SUCCESS, {'tags': []}],
+        ['get', '/api/bookmarks', Response.SUCCESS, {'bookmarks': []}],
+        ['get', '/api/bookmarks/search', Response.SUCCESS, {'bookmarks': []}],
+        ['post', '/api/bookmarks/refresh', Response.FAILURE, None]
     ]
 )
-def test_api_empty_db(client, url, exp_res):
-    if url == '/api/bookmarks/refresh':
-        rd = client.post(url)
-        assert rd.status_code == 400
-    else:
-        rd = client.get(url)
-        assert rd.status_code == 200
-    assert rd.get_json() == exp_res
+def test_api_empty_db(client, method, url, exp_res, data):
+    rd = getattr(client, method)(url)
+    assert_response(rd, exp_res, data)
 
 
 @pytest.mark.parametrize(
-    'url, exp_res, status_code, method', [
-        ['/api/tags/1', {'message': 'This resource does not exist.'}, 404, 'get'],
-        ['/api/tags/1', response_template['failure'], 400, 'put'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'put'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'delete'],
-        ['/api/bookmarks/1/refresh', response_template['failure'], 400, 'post'],
-        ['/api/bookmarks/1/tiny', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1/2', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1/2', response_template['failure'], 400, 'put'],
-        ['/api/bookmarks/1/2', response_template['failure'], 400, 'delete'],
+    'url, methods', [
+        ['api/tags', ['post', 'put', 'delete']],
+        ['/api/tags/tag1', ['post']],
+        ['api/bookmarks', ['put']],
+        ['/api/bookmarks/1', ['post']],
+        ['/api/bookmarks/refresh', ['get', 'put', 'delete']],
+        ['api/bookmarks/1/refresh', ['get', 'put', 'delete']],
+        ['api/bookmarks/1/tiny', ['post', 'put', 'delete']],
+        ['/api/bookmarks/1/2', ['post']],
     ]
 )
-def test_invalid_id(client, url, exp_res, status_code, method):
+def test_not_allowed(client, url, methods):
+    for method in methods:
+        rd = getattr(client, method)(url)
+        assert rd.status_code == HTTP_405_METHOD_NOT_ALLOWED
+
+
+@pytest.mark.parametrize(
+    'method, url', [
+        ['put', '/api/tags/1'],
+        ['post', '/api/bookmarks'],
+        ['put', '/api/bookmarks/1'],
+    ]
+)
+def test_invalid_request(client, method, url):
     rd = getattr(client, method)(url)
-    assert rd.status_code == status_code
-    assert rd.get_json() == exp_res
+    assert rd.status_code in [
+        Response.JSON_NOT_VALID.status_code,
+        HTTP_415_UNSUPPORTED_MEDIA_TYPE
+    ]
+
+
+@pytest.mark.parametrize(
+    'method, url, json, exp_res', [
+        ['get', '/api/tags/tag1', None, Response.TAG_NOT_FOUND],
+        ['put', '/api/tags/tag1', {'tags': ['tag2']}, Response.SUCCESS],
```

Comment:
I am still feeling a bit uneasy about returning SUCCESS when replacing an non-existing tag (i.e. successfully replacing 0 occurrences of a tag). I feel like returning the number of records that were updated  in the response json would be better.

Reply:
The library function doesn't return that information though… May need to get @jarun to approve changing the code API again for that.

Or maybe just return 404 for non-existent tags instead?

Reply:
404 works for me (as suggested in Issue #672).  I would rather open a separate PR if @jarun  later agrees to return the number of updated records.